katdal/lazy_indexer.py

```diff
@@ -443,11 +444,26 @@ def dataset(self):
                 self._orig_dataset = None
             return self._dataset
 
+    def dask_getitem(self, keep):
+        """Index the array and return a dask array.
+
+        This is functionally equivalent to ``self.dataset[keep]``, but it culls
+        unnecessary nodes from the graph, which makes it cheaper to compute if
+        only a small piece of the graph is needed.
+        """
+        # dask does culling anyway as part of optimization, but it first calls
+        # ensure_dict, which copies all the keys, presumably to speed up the
+        # case where most keys are retained. A lazy indexer is normally used to
+        # fetch a small part of the data.
+        kept = self.dataset[keep]
+        kept.dask = dask.optimization.cull(kept.dask, kept.__dask_keys__())[0]
```

Comment:
Does this prevent further optimizations?

Reply:
No, and in fact it'll end up going through the cull process a second time, which is a bit wasteful, but overall it's still a major improvement unless you're passing the whole file through dask.